File: sources/autoscout24.py
"""AutoScout24.ch source — uses the official internal JSON API.

Endpoint reverse-engineered:  POST https://api.autoscout24.ch/v1/listings/search
Filter schema lives under "query":
    makeModelVersions: [{makeKey, modelKey}]
    priceFrom / priceTo
    firstRegistrationYearFrom / firstRegistrationYearTo
    mileageFrom / mileageTo
    cubicCapacityFrom / cubicCapacityTo   (Hubraum in ccm)
    horsePowerFrom / horsePowerTo
    fuelTypes: [petrol|diesel|electric|...]
Sort types: CREATED_DATE, PRICE, MILEAGE, FIRST_REGISTRATION_DATE, HORSE_POWER, RELEVANCE
"""
from __future__ import annotations
import truststore; truststore.inject_into_ssl()   # use Windows cert store (TLS-intercept machine)
import requests
from core import Listing

API = "https://api.autoscout24.ch/v1/listings/search"
HEADERS = {
    "User-Agent": ("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
                   "(KHTML, like Gecko) Chrome/124.0 Safari/537.36"),
    "Accept": "application/json",
    "Accept-Language": "de-CH,de;q=0.9",
    "Content-Type": "application/json",
    "Origin": "https://www.autoscout24.ch",
    "Referer": "https://www.autoscout24.ch/",
}
PAGE_SIZE = 20   # AutoScout24 caps page size at 20


def _build_query(profile) -> dict:
    q = {}
    if profile.autoscout24_models:
        q["makeModelVersions"] = profile.autoscout24_models
    if profile.price_min:
        q["priceFrom"] = profile.price_min
    if profile.price_max < 1_000_000:
        q["priceTo"] = profile.price_max
    if profile.year_min > 1900:
        q["firstRegistrationYearFrom"] = profile.year_min
    if profile.year_max < 2100:
        q["firstRegistrationYearTo"] = profile.year_max
    if profile.mileage_max < 10_000_000:
        q["mileageTo"] = profile.mileage_max
    if profile.cubic_capacity_min:
        q["cubicCapacityFrom"] = profile.cubic_capacity_min
    if profile.horsepower_min:
        q["horsePowerFrom"] = profile.horsepower_min
    if profile.fuel_types:
        q["fuelTypes"] = profile.fuel_types
    return q


def _to_listing(item: dict) -> Listing:
    seller = item.get("seller") or {}
    loc = " ".join(str(x) for x in [seller.get("zipCode", ""), seller.get("city", "")]).strip()
    make = (item.get("make") or {}).get("name", "")
    title = f"{make} {item.get('versionFullName', '')}".strip()
    lid = str(item.get("id"))
    images = item.get("images") or []
    img = ""
    if images:
        first = images[0]
        img = first if isinstance(first, str) else (first.get("url") or first.get("key") or "")
    return Listing(
        source="autoscout24",
        listing_id=lid,
        url=f"https://www.autoscout24.ch/de/d/{lid}",
        title=title or f"Listing {lid}",
        price=int(item["price"]) if item.get("price") is not None else None,
        year=item.get("firstRegistrationYear"),
        mileage=item.get("mileage"),
        horsepower=item.get("horsePower"),
        fuel=item.get("fuelType"),
        location=loc,
        image=img,
        raw=item,
    )


def fetch(profile, max_listings: int = 250) -> list[Listing]:
    """Return the cheapest `max_listings` current listings matching the profile.

    Cheapest-first (PRICE ASC) keeps the most deal-relevant cars even when the
    market is large, and gives a clean price distribution for deal scoring.
    """
    query = _build_query(profile)
    out: list[Listing] = []
    session = requests.Session()
    session.headers.update(HEADERS)
    page = 0
    while len(out) < max_listings:
        body = {
            "query": query,
            "sort": [{"type": "PRICE", "order": "ASC"}],
            "pagination": {"page": page, "size": PAGE_SIZE},
        }
        r = session.post(API, json=body, timeout=30)
        if r.status_code != 200:
            raise RuntimeError(f"AutoScout24 API {r.status_code}: {r.text[:200]}")
        data = r.json()
        items = data.get("content") or []
        if not items:
            break
        out.extend(_to_listing(it) for it in items)
        if data.get("last") or page + 1 >= data.get("totalPages", 1):
            break
        page += 1
    return out[:max_listings]

File: cf-worker/src/autoscout24.js
// AutoScout24.ch client — internal JSON API (no browser needed).
// POST https://api.autoscout24.ch/v1/listings/search  with filters under "query".

const API = "https://api.autoscout24.ch/v1/listings/search";
const MAKES = "https://api.autoscout24.ch/v1/makes?vehicleCategory=car";
const HEADERS = {
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
    "(KHTML, like Gecko) Chrome/124.0 Safari/537.36",
  "Accept": "application/json",
  "Accept-Language": "de-CH,de;q=0.9",
  "Content-Type": "application/json",
  "Origin": "https://www.autoscout24.ch",
  "Referer": "https://www.autoscout24.ch/",
};
const PAGE_SIZE = 20; // API caps page size at 20

function buildQuery(search) {
  const q = {};
  if (search.makeKey) {
    const mm = { makeKey: search.makeKey };
    if (search.modelKey) mm.modelKey = search.modelKey;
    if (search.modelGroupKey) mm.modelGroupKey = search.modelGroupKey;  // Baureihe, z. B. "3-series"
    q.makeModelVersions = [mm];
  }
  if (search.bodyTypes && search.bodyTypes.length) q.bodyTypes = search.bodyTypes;
  if (search.maxPrice) q.priceTo = search.maxPrice;
  if (search.minPrice) q.priceFrom = search.minPrice;
  if (search.minYear) q.firstRegistrationYearFrom = search.minYear;
  if (search.maxYear) q.firstRegistrationYearTo = search.maxYear;
  if (search.maxMileage) q.mileageTo = search.maxMileage;
  if (search.minCcm) q.cubicCapacityFrom = search.minCcm;
  if (search.minHp) q.horsePowerFrom = search.minHp;
  if (search.fuelTypes && search.fuelTypes.length) q.fuelTypes = search.fuelTypes;
  return q;
}

const IMG_BASE = "https://listing-images.autoscout24.ch/";

function imageUrl(item) {
  const imgs = item.images || [];
  if (!imgs.length) return "";
  const first = imgs[0];
  const key = typeof first === "string" ? first : first.key || first.url || "";
  if (!key) return "";
  return key.startsWith("http") ? key : IMG_BASE + key;
}

function toListing(item) {
  const seller = item.seller || {};
  const loc = [seller.zipCode || "", seller.city || ""].join(" ").trim();
  const make = (item.make || {}).name || "";
  const id = String(item.id);
  return {
    source: "autoscout24",
    uid: "autoscout24:" + id,
    url: "https://www.autoscout24.ch/de/d/" + id,
    title: (make + " " + (item.versionFullName || "")).trim() || "Listing " + id,
    price: item.price != null ? Math.round(item.price) : null,
    year: item.firstRegistrationYear || null,
    mileage: item.mileage != null ? item.mileage : null,
    horsepower: item.horsePower || null,
    fuel: item.fuelType || null,
    location: loc,
    teaser: item.teaser || "",
    image: imageUrl(item),
    created: item.createdDate || null,
  };
}

// AS24-Listing-IDs sind rein numerisch. Hart darauf festnageln, damit nichts
// anderes (z. B. ein manipuliertes uid) in den API-Pfad gelangen kann.
function safeListingId(id) {
  const s = String(id);
  return /^\d+$/.test(s) ? s : null;
}

// Full listing description (for keyword filtering on the ad text).
export async function fetchDescription(id) {
  const sid = safeListingId(id);
  if (!sid) return "";
  try {
    const r = await fetch("https://api.autoscout24.ch/v1/listings/" + sid, { headers: HEADERS });
    if (!r.ok) return "";
    const d = await r.json();
    return [d.description || "", d.teaser || ""].join(" ");
  } catch (e) {
    return "";
  }
}

// Alle Bild-URLs eines Detail-Items (für die /mehrinfo-Galerie).
function allImageUrls(item) {
  const imgs = item.images || [];
  const out = [];
  for (const im of imgs) {
    const key = typeof im === "string" ? im : im.key || im.url || "";
    if (!key) continue;
    out.push(key.startsWith("http") ? key : IMG_BASE + key);
  }
  return out;
}

// Vollständiges Inserat (für /mehrinfo). Liefert das rohe Detail-Objekt + alle
// Bild-URLs. Null bei Fehler/Nicht-gefunden.
export async function fetchListingDetail(id) {
  const sid = safeListingId(id);
  if (!sid) return null;
  try {
    const r = await fetch("https://api.autoscout24.ch/v1/listings/" + sid, { headers: HEADERS });
    if (!r.ok) return null;
    const d = await r.json();
    d._images = allImageUrls(d);
    return d;
  } catch (e) {
    return null;
  }
}

// Fetch up to maxListings cheapest matching listings (PRICE ASC).
export async function fetchListings(search, maxListings = 250) {
  const query = buildQuery(search);
  const out = [];
  let page = 0;
  while (out.length < maxListings) {
    const body = {
      query,
      sort: [{ type: "PRICE", order: "ASC" }],
      pagination: { page, size: PAGE_SIZE },
    };
    const r = await fetch(API, { method: "POST", headers: HEADERS, body: JSON.stringify(body) });
    if (!r.ok) throw new Error("AS24 " + r.status + ": " + (await r.text()).slice(0, 200));
    const data = await r.json();
    const items = data.content || [];
    if (!items.length) break;
    for (const it of items) out.push(toListing(it));
    if (data.last || page + 1 >= (data.totalPages || 1)) break;
    page++;
  }
  return out.slice(0, maxListings);
}

// All makes (key + name), for validating typed brands / matching.
export async function fetchMakes() {
  const r = await fetch(MAKES, { headers: HEADERS });
  if (!r.ok) throw new Error("AS24 makes " + r.status);
  const d = await r.json();
  const items = Array.isArray(d) ? d : d.content || [];
  return items.map((m) => ({ key: m.key, name: m.name }));
}

// Models of a make (key + name + group), for matching typed model names.
export async function fetchModels(makeKey) {
  const url = "https://api.autoscout24.ch/v1/makes/key/" + encodeURIComponent(makeKey) +
    "/models?vehicleCategory=car";
  const r = await fetch(url, { headers: HEADERS });
  if (!r.ok) throw new Error("AS24 models " + r.status);
  const d = await r.json();
  const items = Array.isArray(d) ? d : d.content || [];
  return items.map((m) => ({
    key: m.key,
    name: m.name,
    // Baureihe als Objekt {key,name} — z. B. {key:"3-series", name:"3 SERIES"}.
    group: m.group && m.group.key ? { key: m.group.key, name: m.group.name || "" } : null,
  }));
}
